File: aemet/SergioDeIscar-aemet-java/SergioDeIscar-aemet-java/src/main/java/org/example/services/storage/informe/InformeFileJson.java
package org.example.services.storage.informe;

import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.Moshi;
import org.example.config.AppConfig;
import org.example.dto.InformeDto;
import org.example.dto.InformesDto;
import org.example.models.Informe;

import java.io.*;
import java.nio.file.Files;
import java.util.List;

public class InformeFileJson implements InformeStorageService {
    //Singleton
    private static InformeFileJson instance = null;
    public static InformeFileJson getInstance() {
        if (instance == null) instance = new InformeFileJson();
        return instance;
    }

    private final String localFile = AppConfig.getInstance().getAppData() + File.separator + "informes.json";
    private final Moshi moshi = new Moshi.Builder().build();

    @Override
    public List<Informe> saveAll(List<Informe> elements) {
        File file = new File(localFile);
        if (file.exists() && !file.canWrite() ) return null;
        JsonAdapter<InformesDto> jsonAdapter = moshi.adapter(InformesDto.class);
        try(PrintWriter pw = new PrintWriter(new FileWriter(file))) {
            pw.println(
                    jsonAdapter.indent("\t").toJson(
                            new InformesDto(
                                    elements.stream().map(Informe::toDto).toList()
                            )
                    )
            );
            return elements;
        } catch (IOException e) {
            System.out.println("Error al escribir en el fichero json");
        }
        return null;
    }

    @Override
    public List<Informe> loadAll() {
        File file = new File(localFile);
        if(!file.exists() || !file.canRead()) return null;
        JsonAdapter<InformesDto> jsonAdapter = moshi.adapter(InformesDto.class);
        try {
            String json = Files.readString(file.toPath());
            InformesDto dto = jsonAdapter.fromJson(json);
            return dto.getInformes().stream().map(InformeDto::toClass).toList();

        } catch (IOException e) {
            System.out.println("Error al leer en el fichero json");
        }
        return null;
    }
}

File: aemet/SergioDeIscar-aemet-kotlin/SergioDeIscar-aemet-kotlin/src/main/kotlin/services/storage/informe/InformeFileJson.kt
package services.storage.informe

import com.squareup.moshi.Moshi
import com.squareup.moshi.adapter
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import config.AppConfig
import dto.InformeDto
import mappers.toClass
import mappers.toDto
import models.Informe
import mu.KotlinLogging
import java.io.File

private val logger = KotlinLogging.logger {  }

@ExperimentalStdlibApi
object InformeFileJson: InformeStorageService{
    private val localFile = "${AppConfig.APP_DATA}${File.separator}informe.json"
    private val moshi = Moshi.Builder()
        .add(KotlinJsonAdapterFactory())
        .build()

    override fun saveAll(elements: List<Informe>): List<Informe> {
        logger.debug { "InformeFileJson ->\tsaveAll" }

        val file = File(localFile)
        if (file.exists() && !file.canWrite()) return emptyList()

        val jsonAdapter = moshi.adapter<List<InformeDto>>()
        val json = jsonAdapter.indent("\t").toJson(elements.map { it.toDto() })

        file.writeText(json)

        return elements
    }

    override fun loadAll(): List<Informe> {
        logger.debug { "InformeFileJson ->\tloadAll" }

        val file = File(localFile)
        if (!file.exists() || !file.canRead()) return emptyList()

        val jsonAdapter = moshi.adapter<List<InformeDto>>()

        return jsonAdapter.fromJson(file.readText())?.map { it.toClass() } ?: emptyList()
    }
}